app/src/main/java/com/example/android/codelabs/paging/ui/SearchRepositoriesViewModel.kt

```diff
@@ -41,20 +40,48 @@ class SearchRepositoriesViewModel(private val repository: GithubRepository) : Vi
     @Volatile
     private var currentQueryValue: String? = null
     @Volatile
-    private var currentSearchResult: Flow<PagingData<Repo>>? = null
+    private var currentSearchResult: Flow<PagingData<UiModel>>? = null
 
     /**
      * Search a repository based on a query string.
      */
-    fun searchRepo(queryString: String): Flow<PagingData<Repo>> {
+    fun searchRepo(queryString: String): Flow<PagingData<UiModel>> {
         val lastResult = currentSearchResult
         if (queryString == currentQueryValue && lastResult != null) {
             return lastResult
         }
         currentQueryValue = queryString
-        val newResult: Flow<PagingData<Repo>> = repository.getSearchResultStream(queryString)
+        val newResult: Flow<PagingData<UiModel>> = repository.getSearchResultStream(queryString)
+                .map { pagingData -> pagingData.map { UiModel.RepoItem(it) } }
+                .map {
+                    it.insertSeparators<UiModel.RepoItem, UiModel> { before, after ->
+                        if (before == null && after is UiModel.RepoItem) {
+                            Log.d("ViewModel", "-> adding null separator")
+                            if (after.repo.stars >= 10_000) {
+                                UiModel.SeparatorItem("${after.repo.stars / 10_000}0.000+ stars")
+                            } else {
+                                UiModel.SeparatorItem("< 10.000+ stars")
+                            }
+                        } else if (before is UiModel.RepoItem && after is UiModel.RepoItem
+                                && before.repo.stars / 10_000 > after.repo.stars / 10_000) {
+                            if (after.repo.stars >= 10_000) {
+                                UiModel.SeparatorItem("${after.repo.stars / 10_000}0.000+ stars")
+                            } else {
+                                UiModel.SeparatorItem("< 10.000+ stars")
+                            }
+                        } else {
+                            // no separator
+                            null
+                        }
+                    }
+                }
```

Comment:
suggest: Something along these lines seems a bit easier to read to me than the nested ifs
```
fun UiModel.starCount(): Int {
    return ((before as? UiModel.RepoItem)?.repo?.stars ?: 0) / 10_000
}

fun Int.toStarString: String = when {
    this > 0 -> "$this0.000+ stars"
    else -> "< 10.000+ stars"
}

val beforeStars = before.starCount()
val afterStars = after.starCount()

when {
    after == null -> null
    before == null || beforeStars > afterStars -> afterStars.toStarString()
    else -> null
}
```

Reply:
I like the direction, but I'd put both functions as methods on UiModel.RepoItem, since that cast is unnecessary, and this is Ui behavior the UiModel can do.

(Note that the cast to RepoItem should no longer be necessary, as mentioned in prev comment)

```
// extension to get rounded star count (or could go in class, since this is UI functionality)
val UiModel.RepoItem.roundedStarCount: Int = this.repo.starCount / 10_000) * 10_000
```


I'm wondering again if we should have an extension for the following behavior, since it's what you want 99% of the time:
```
if (after != null &&
        (before == null || before.roundedStarCount != after.roundedStarCount)) {
    Separator(before.roundedStarCountString)
} else {
    null
}
```

It would mean yet another generic on insertSeparators though :P

Reply:
`roundedStarCount` it is